hc-nested-list-behavior.html

```diff
@@ -0,0 +1,178 @@
+       *  when lazy-loading data add a new key-value pair to this object containing the data.  The
+       *  key value is used for looking up parentId to fetch the list of child items.  All item data
+       *  is taken from the items array.
+       */
+      parts: {
+        type: Object,
+        notify: true
+      }
+    },
+    observers: ['_computeParts(items)'],
+    listeners: {
+      toggle: 'toggleOpen'
+    },
+    /**
+     * Takes a parts object and optionally a start point for an array then iterates over the
+     *  object starting at the root if there isn't a location already set and finds all open nodes
+     *  recursively calling itself to ensure order is preserved once flattened.
+     *
+     * @param {Object} parts
+     * @param {Object[]} [items=parts.root]
+     * @param {String} [idName=this.itemIdName]
+     * @returns {Object[]}
+     */
+    _computeItems(parts, items = parts.root, idName = this.itemIdName) {
+      // Can't calculate items without parts or the name of the id property.
+      if (!idName || !parts) return [];
+
+      const newItems = [];
+
+      // Push all items into the new array, and if there is a child item then push all of the children.
+      items.forEach(item => {
+        item.open = Boolean(item.open);
+        newItems.push(item);
+        if (item.open)
+          newItems.push(...this._computeItems(parts, parts[item[idName]] || []));
+      });
+
+      return newItems;
+    },
+    /**
+     * Takes the flat items structure and builds out a map of parent-child elements which gets
+     *  stored and used for optimized rendering later.
+     *
+     * @param {Object[]} items
+     */
+    _computeParts(items) {
+      const parts = {};
+
+      items.forEach(item => {
+        // Fetch the parentId or set it to 'root' if there isn't one
+        const parentId = this._getParentId(item);
+        if (!parts[parentId]) parts[parentId] = [];
+        parts[parentId].push(item);
+      });
+
+      this.set('parts', parts);
+    },
+    /**
+     * Used to fetch the parentId if it's available or default back to 'root'.
+     *
+     * @param {Object} item
+     * @param {String} [parentIdName=this.itemParentIdName]
+     * @returns {String|Number}
+     */
+    _getParentId(item, parentIdName = this.itemParentIdName) {
+      return (Boolean(item[parentIdName]) || item[parentIdName] === 0) ?
+        item[parentIdName] :
+        'root';
+    },
+    /**
+     * Called when an item's open status is changed.  Update parts with children if lazy-loading then call update
+     *  with the updated parts object once complete, update will always need to be called even if no changes were
+     *  made to the parts object.  It's recommended to optimize your loader by checking to see if items already exist
+     *  before attempting to fetch them.
+     *
+     * @param {Object} item
+     * @param {Object} parts
+     * @param {Function} update
+     */
+    fetchChildren(item, parts, update) {
```

Comment:
This function is designed for the lazy-loading process, in order to add your own custom logic for lazy-loading you can overwrite this function, which will require the item in order to fetch the proper data, as well as the parts object which you need to update then pass to the update function.